## Exploring eBay Car Sales Data
***
In this project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.
### Dataset Details

Data dictionary for the dataset:

- `dateCrawled` - When the ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing.
- `price` - The listed selling price of the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The type of vehicle.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The type of transmission.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `odometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.

The dataset was originally scraped and uploaded to Kaggle by user orgesleka.
The aim of this project is to clean the dataset and perform some initial analysis on it.


Content
___
* [Cleaning Column Names](#cleaning_names)
* [Initial Exploration and Cleaning](#initial)
* [Exploring the Odometer and Price Columns](#odom_price)
* [Exploring the Date Columns](#date)
* [Exploring Price by Brand](#pbb)
* [Exploring Price by Brand for Entire Dataset ](#pbb_alls)
* [Exploring Mileage ](#miles)
* [Replace German Words](#de_to_eng)
* [Convert the Dates to  Uniform Numeric Data](#uniform)
* [The Most Common Brand/Model Combinations](#common_b_m)
* [The Mileage/Price Analysis](#miles_price) 
* [The Price/Damage Correlation](#damage)

In [1]:
import pandas as pd

In [2]:
autos = pd.read_csv('autos.csv',encoding='latin1')
autos


dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [4]:
autos.head(3)

dateCrawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offerType   price   abtest vehicleType  yearOfRegistration    gearbox  \
0   Angebot  $5,000  control         bus                2004    manuell   
1   Angebot  $8,500  control   limousine                1997  automatik   
2   Angebot  $8,990     test   limousine                2009    manuell   

   powerPS   model   odometer  monthOfRegistration fuelType       brand  \
0      158  andere  150,000km                    3      lpg     peugeot   
1      286     7er  150,000km                    6   benzin         bmw   
2      102    golf   70,000km                    7   benzin  volkswagen   

  notRepairedDamage          dateCreated  nrOfPictures  postalCode  \
0              nein  2016-03-26 00:00:00             0       79588   
1              nein  2016-04-04 00:00:00             0       71034   
2              nein  2016-03-26 00:00:00             0       35394   

              lastSeen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

The dataset contains 20 columns, most of which are strings.
Some columns have null values, but none have more than ~20% null values.



<a id='cleaning_names'></a>
### Cleaning Column Names
The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.
> Next steps will be: 
>* convert the column names from camelcase to snakecase
>* change some of the column names based on the data dictionary in order to be more descriptive

In [5]:
#printing the column names
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

Renaming the columns using snake_case and improve clarity based on the data dictionary.

In [6]:
# Create a copy of the original dataset
clean_col = autos.copy()
# clean_col.columns
clean_col.rename(columns= {'yearOfRegistration': 'registration_year', 'monthOfRegistration': 'registration_month','notRepairedDamage': 'unrepaired_damage', 'dateCreated': 'ad_created',
                         'dateCrawled': 'date_crawled', 'offerType': 'offer_type', 'vehicleType': 'vehicle_type',
                          'powerPS' : 'power_ps', 'fuelType':'fuel_type', 'nrOfPictures': 'pictures_number',
                          'postalCode': 'postal_code', 'lastSeen': 'last_seen' }, inplace=True)
clean_col.columns


Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'pictures_number', 'postal_code',
       'last_seen'],
      dtype='object')

In [7]:
#asssigning the renamed columns to the original dataset
autos.columns = clean_col.columns
print("The modified columns:")
autos.columns

The modified columns:


Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'pictures_number', 'postal_code',
       'last_seen'],
      dtype='object')

<a id='initial'></a>
### Initial Exploration and Cleaning
Now let's do some basic data exploration to determine what other cleaning tasks need to be done.

Initially we will look for:

* Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
* Columns like __odometer__ and  __price__ should be converted into numerical datatype.

In [8]:
# Get both categorical and numeric columns)
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  pictures_number   postal_code  \
count                 50000          50000.0  50000.000000   
unique                   76              NaN           NaN   
top     2016-04-03 00:00:00              NaN           NaN   
freq                   1946              NaN           NaN   
mean                    NaN              0.0  50813.627300   
std                     NaN              0.0  25779.747957   
min                     NaN              0.0   1067.000000   
25%                     NaN              0.0  30451.000000   
50%                     NaN              0.0  49577.000000   
75%                     NaN              0.0  71540.000000   
max                     NaN              0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [9]:
autos['price'].value_counts()

$0         1421
$500        781
$1,500      734
$2,500      643
$1,000      639
           ... 
$414          1
$79,933       1
$5,198        1
$18,890       1
$16,995       1
Name: price, Length: 2357, dtype: int64

In [10]:
autos.isnull().sum()

date_crawled             0
name                     0
seller                   0
offer_type               0
price                    0
abtest                   0
vehicle_type          5095
registration_year        0
gearbox               2680
power_ps                 0
model                 2758
odometer                 0
registration_month       0
fuel_type             4482
brand                    0
unrepaired_damage     9829
ad_created               0
pictures_number          0
postal_code              0
last_seen                0
dtype: int64


The data in columns `vehicle_type ` ,`gearbox`,`model`,`fuel_type`, `unrepaired_damage` has missing values.

In [11]:
print(f'The total number of missing values: {autos.isnull().sum().sum()}')

The total number of missing values: 24844


### Findings:
1. For the **price** column: Some prices are zeros, and not numerical.
2. Gearbox, odometer, car power columns are not numerical.
3. Missing values: *Vehicle type* is missing  5095 values, *gearbox*: 2680, *unrepaired_damage*: 9829,*fuel type*: 4482,*model*: 2758
4. *Seller* and *offer type* columns have the same(or nearly all) values.
5. All values in number of pictures are 0, so let`s drop thic column, as well as 'offer type' and 'seller'

In [12]:
#Exploring number of pictures
autos["pictures_number"].value_counts()

0    50000
Name: pictures_number, dtype: int64

In [13]:
#Dropping columns
autos = autos.drop(['pictures_number', 'seller', 'offer_type'], axis=1)

The dataset includes entries with a price of zero, which may require investigation.

In [14]:
#Created a df that contains cars with prices set to zero.
zero_price = autos[autos['price'] == '$0']
zero_price


date_crawled                                               name  \
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
71     2016-03-28 19:39:35  Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...   
80     2016-03-09 15:57:57  Nissan_Primera_Hatchback_1_6_16v_73_Kw___99Ps_...   
87     2016-03-29 23:37:22                      Bmw_520_e39_zum_ausschlachten   
99     2016-04-05 09:48:54                    Peugeot_207_CC___Cabrio_Bj_2011   
...                    ...                                                ...   
49884  2016-03-11 13:55:30               Audi_a6_2.5l__Schnaeppchen_nur_heute   
49943  2016-03-16 20:46:08                                         Opel_astra   
49960  2016-03-25 22:51:55            Ford_KA_zu_verschenken_***Reserviert***   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   

      price   abtest vehicle_type  registration_year  gearbox  power_ps  \
27       $0  control          NaN               2005      NaN         0   
71       $0  control          NaN               1990  manuell         0   
80       $0  control        coupe               1999  manuell        99   
87       $0  control          NaN               2000      NaN         0   
99       $0  control       cabrio               2011  manuell         0   
...     ...      ...          ...                ...      ...       ...   
49884    $0     test        kombi               1999  manuell       150   
49943    $0  control          NaN               2016  manuell       101   
49960    $0  control   kleinwagen               1999  manuell        60   
49974    $0  control       cabrio               1983  manuell        70   
49984    $0     test          NaN               2000      NaN         0   

         model   odometer  registration_month fuel_type           brand  \
27         NaN  150,000km                   0       NaN            ford   
71         NaN    5,000km                   0    benzin            opel   
80     primera  150,000km                   3    benzin          nissan   
87         5er  150,000km                   0       NaN             bmw   
99     2_reihe   60,000km                   7    diesel         peugeot   
...        ...        ...                 ...       ...             ...   
49884       a6  150,000km                  11    diesel            audi   
49943    astra  150,000km                   8    benzin            opel   
49960       ka  150,000km                   6    benzin            ford   
49974     golf  150,000km                   2    benzin      volkswagen   
49984      NaN  150,000km                   0       NaN  sonstige_autos   

      unrepaired_damage           ad_created  postal_code            last_seen  
27                  NaN  2016-03-27 00:00:00        66701  2016-03-27 18:45:01  
71                  NaN  2016-03-28 00:00:00         4552  2016-04-07 01:45:48  
80                   ja  2016-03-09 00:00:00        66903  2016-03-09 16:43:50  
87                  NaN  2016-03-29 00:00:00        82256  2016-04-06 21:18:15  
99                 nein  2016-04-05 00:00:00        99735  2016-04-07 12:17:34  
...                 ...                  ...          ...                  ...  
49884               NaN  2016-03-11 00:00:00        27711  2016-03-12 03:17:08  
49943               NaN  2016-03-16 00:00:00        89134  2016-03-17 19:44:20  
49960               NaN  2016-03-25 00:00:00        34355  2016-03-25 22:51:55  
49974              nein  2016-03-20 00:00:00         8209  2016-03-27 19:48:16  
49984               NaN  2016-03-31 00:00:00        12103  2016-04-02 19:44:53  

[1421 rows x 17 columns]

The price and odometer columns are numeric values stored as text. 

For each column:
- Remove any non-numeric characters.
- Convert the column to a numeric dtype.
Also, we will rename the *odometer* column to *odometer_km* for convenience.

In [15]:
#Rename odometer to odometer_km
autos.rename(columns={'odometer': 'odometer_km'}, inplace=True)
autos.head(3)

date_crawled                                        name   price  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  $5,000   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  $8,500   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  $8,990   

    abtest vehicle_type  registration_year    gearbox  power_ps   model  \
0  control          bus               2004    manuell       158  andere   
1  control    limousine               1997  automatik       286     7er   
2     test    limousine               2009    manuell       102    golf   

  odometer_km  registration_month fuel_type       brand unrepaired_damage  \
0   150,000km                   3       lpg     peugeot              nein   
1   150,000km                   6    benzin         bmw              nein   
2    70,000km                   7    benzin  volkswagen              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2  2016-03-26 00:00:00        35394  2016-04-06 20:15:37

In [16]:
#Remove non numerical characters from "price" and convert to numerical data type.
autos['price'] = autos['price'].str.replace('$', '', regex=False).str.replace(',', '', regex=False)
autos['price'] = autos['price'].astype('float')
autos['price'].head()

0    5000.0
1    8500.0
2    8990.0
3    4350.0
4    1350.0
Name: price, dtype: float64

In [17]:
#Cleaning 'odometer_km' column by removing 'km' and converting to numerical datatype.
autos['odometer_km'] = autos['odometer_km'].str.replace('km', '', regex=False)
autos['odometer_km'] = autos['odometer_km'].str.replace(',', '', regex=False)
#Convert 'odometer_km' to numerical format.
autos['odometer_km'] = autos['odometer_km'].astype('float')
autos['odometer_km'].head(5)

0    150000.0
1    150000.0
2     70000.0
3     70000.0
4    150000.0
Name: odometer_km, dtype: float64

The modified data set:

In [18]:
autos.head(3)


date_crawled                                        name   price  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  5000.0   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  8500.0   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  8990.0   

    abtest vehicle_type  registration_year    gearbox  power_ps   model  \
0  control          bus               2004    manuell       158  andere   
1  control    limousine               1997  automatik       286     7er   
2     test    limousine               2009    manuell       102    golf   

   odometer_km  registration_month fuel_type       brand unrepaired_damage  \
0     150000.0                   3       lpg     peugeot              nein   
1     150000.0                   6    benzin         bmw              nein   
2      70000.0                   7    benzin  volkswagen              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2  2016-03-26 00:00:00        35394  2016-04-06 20:15:37

<a id='odom_price'></a>
### Exploring the Odometer and Price Columns
#### Odometer

In [19]:
#max and min values
print(f'There are: {len(autos["odometer_km"].unique())} unique values.')

There are: 13 unique values.


In [20]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [21]:
autos["odometer_km"].value_counts()

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64

As we can see, the majority of cars have mileage at least 150,000 km.

#### Price

In [22]:
autos["price"].unique().shape

(2357,)

In [23]:
print(autos["price"].describe())

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


In [24]:
autos["price"].unique().shape


(2357,)

In [25]:
print(f'The max price: {autos["price"].max()}')

The max price: 99999999.0


In [26]:
autos["price"].sort_index(ascending=True).head(20)

0      5000.0
1      8500.0
2      8990.0
3      4350.0
4      1350.0
5      7900.0
6       300.0
7      1990.0
8       250.0
9       590.0
10      999.0
11      350.0
12     5299.0
13     1350.0
14     3999.0
15    18900.0
16      350.0
17     5500.0
18      300.0
19     4150.0
Name: price, dtype: float64

Given that eBay is an auction site, there could legitimately be items where the opening bid is 
1 dollar,but remove anything above $350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

In [27]:
#Selecting all data with prices in range from $1 to 351,000
autos = autos[autos["price"].between(1,351000)]
autos.head(5)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0  5000.0  control          bus               2004    manuell       158   
1  8500.0  control    limousine               1997  automatik       286   
2  8990.0     test    limousine               2009    manuell       102   
3  4350.0  control   kleinwagen               2007  automatik        71   
4  1350.0     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere     150000.0                   3       lpg     peugeot   
1     7er     150000.0                   6    benzin         bmw   
2    golf      70000.0                   7    benzin  volkswagen   
3  fortwo      70000.0                   6    benzin       smart   
4   focus     150000.0                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

<a id='date'></a>
### Exploring the Date Columns
There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself.
The date_crawled, last_seen, and ad_created columns are all identified as string values by pandas.

In [28]:
#The data type of the date columns
autos[['date_crawled','last_seen','ad_created']].dtypes

date_crawled    object
last_seen       object
ad_created      object
dtype: object

The dates that came from the website are  in the numerical format. 

In [29]:
autos[['registration_month','registration_year']].dtypes

registration_month    int64
registration_year     int64
dtype: object

These columns all represent full timestamp values:

In [30]:
autos[['date_crawled','last_seen','ad_created']][0:5]

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

In [31]:
print(autos['date_crawled'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 48565, dtype: object


To understand the date range, we can extract just the date values, use Series.value_counts() to generate a distribution, and then sort by the index.

In [32]:
autos['date_crawled'].str[:10].value_counts()

2016-04-03    1875
2016-03-20    1840
2016-03-21    1815
2016-03-12    1793
2016-03-14    1775
2016-04-04    1772
2016-03-07    1749
2016-04-02    1723
2016-03-28    1693
2016-03-19    1689
2016-03-15    1665
2016-03-29    1656
2016-03-30    1636
2016-04-01    1636
2016-03-08    1617
2016-03-09    1607
2016-03-22    1602
2016-03-11    1582
2016-03-23    1565
2016-03-26    1564
2016-03-10    1563
2016-03-31    1546
2016-03-17    1536
2016-03-25    1535
2016-03-27    1510
2016-03-16    1438
2016-03-24    1425
2016-03-05    1230
2016-03-13     761
2016-03-06     682
2016-04-05     636
2016-03-18     627
2016-04-06     154
2016-04-07      68
Name: date_crawled, dtype: int64

In [33]:
autos['date_crawled'].str[:10].sort_index(ascending=False)

49999    2016-03-14
49998    2016-03-08
49997    2016-04-02
49996    2016-03-28
49995    2016-03-27
            ...    
4        2016-04-01
3        2016-03-12
2        2016-03-26
1        2016-04-04
0        2016-03-26
Name: date_crawled, Length: 48565, dtype: object

In [34]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_values()

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-04-01    0.033687
2016-03-30    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.





In [35]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64


The last three days contain a disproportionate amount of 'last seen' values. 
- 2016-04-05    **0.124761**
- 2016-04-06   **0.221806**
- 2016-04-07   **0.131947**

Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [36]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

In [37]:
print(f"The number of created ads: {autos['ad_created'].value_counts().sum()}.")



The number of created ads: 48565.


There is a large variety of ad created dates(48565). Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

In [38]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The registartion year is the year in which the car was first registered, but it has odd dates:
*  the minimum year is 1000  
*  the maximum year is 9999

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely.

In [39]:
autos['registration_year']

0        2004
1        1997
2        2009
3        2007
4        2003
         ... 
49995    2011
49996    1996
49997    2014
49998    2013
49999    1996
Name: registration_year, Length: 48565, dtype: int64

In [40]:
# counts the number of vehicles in the autos  that have registration years outside the range of 1900 to 2016 (inclusive).
filtered_data = (~autos["registration_year"].between(1900,2016)).sum() 
print(f"The number of cars outside the 1900-2016 years range: {filtered_data}.")

The number of cars outside the 1900-2016 years range: 1884.


In [41]:
#number of rows
autos.shape[0]

48565

In [42]:
percentage_outside = (filtered_data / autos.shape[0]) * 100
print(f"The percentage of cars outside the 1900-2016 range: {percentage_outside}.")

The percentage of cars outside the 1900-2016 range: 3.8793369710697.


In [43]:
# Get the distribution of registration years (normalized to percentages)
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.064985
2005    0.060455
1999    0.059652
2004    0.055657
2003    0.055575
2006    0.054978
2001    0.054278
2002    0.051189
1998    0.048656
2007    0.046886
Name: registration_year, dtype: float64

It appears that most of the vehicles were first registered in the past 20 years.

<a id='pbb'></a>
### Exploring Price by Brand

In [44]:
autos['brand'].value_counts()

volkswagen        10336
opel               5277
bmw                5274
mercedes_benz      4652
audi               4168
ford               3382
renault            2325
peugeot            1430
fiat               1262
seat                920
skoda               780
nissan              741
mazda               739
smart               694
citroen             686
toyota              611
hyundai             483
sonstige_autos      471
volvo               439
mini                418
mitsubishi          399
honda               388
kia                 345
alfa_romeo          321
porsche             287
suzuki              286
chevrolet           275
chrysler            169
dacia               129
daihatsu            122
jeep                108
subaru              103
land_rover           99
saab                 79
daewoo               76
jaguar               74
trabant              68
rover                65
lancia               55
lada                 29
Name: brand, dtype: int64

To analyze prices by brand, we'll first identify the top 20 most frequently occurring brands. However, it's worth noting that there's a brand labeled as "_sonstige_autos_," which essentially represents "other" or "miscellaneous" and encompasses a variety of different brands. For the purose of our analysis, we'll exclude "_sonstige_autos_" and focus solely on the remaining existing brands.

In [45]:
# Dropping 'sonstige_autos' row because it containes "other cars" brands
mask = autos['brand'] == 'sonstige_autos'
autos = autos[~mask]
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

         price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0       5000.0  control          bus               2004    manuell       158   
1       8500.0  control    limousine               1997  automatik       286   
2       8990.0     test    limousine               2009    manuell       102   
3       4350.0  control   kleinwagen               2007  automatik        71   
4       1350.0     test        kombi               2003    manuell         0   
...        ...      ...          ...                ...        ...       ...   
49995  24900.0  control    limousine               2011  automatik       239   
49996   1980.0  control       cabrio               1996    manuell        75   
49997  13200.0     test       cabrio               2014  automatik        69   
49998  22900.0  control        kombi               2013    manuell       150   
49999   1250.0  control    limousine               1996    manuell       101   

        model  odometer_km  registration_month fuel_type       brand  \
0      andere     150000.0                   3       lpg     peugeot   
1         7er     150000.0                   6    benzin         bmw   
2        golf      70000.0                   7    benzin  volkswagen   
3      fortwo      70000.0                   6    benzin       smart   
4       focus     150000.0                   7    benzin        ford   
...       ...          ...                 ...       ...         ...   
49995      q5     100000.0                   1    diesel        audi   
49996   astra     150000.0                   5    benzin        opel   
49997     500       5000.0                  11    benzin        fiat   
49998      a3      40000.0                  11    diesel        audi   
49999  vectra     150000.0                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                  nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                  nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                  nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
...                 ...                  ...          ...                  ...  
49995              nein  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49997              nein  2016-04-02 00:00:00        73430  2016-04-04 11:47:27  
49998              nein  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[48094 rows x 17 columns]

In [46]:
#Selecting the top 20 brands. The most frequetly occuring bands
top_20_brands = autos['brand'].value_counts().nlargest(20).index
top_20_brands

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'volvo', 'mini', 'mitsubishi'],
      dtype='object')

Generate a dictionary comprising the top 20 brands along with their corresponding average prices.

In [47]:
brand_mean_price = {}
for brand in top_20_brands:
    mean_price = autos.loc[autos['brand'] == brand, 'price'].mean()
    brand_mean_price [brand] = int(mean_price)
# print(brand_mean_price )

#Sorting prices
sorted_price_brand = sorted(brand_mean_price .items(), key=lambda x: x[1], reverse=True)
print("\nAnalysis of Average Prices by Brand:")
for brand, mean_price in sorted_price_brand:
    print(f"The average price of {brand} vehicles is ${mean_price:.2f}")


Analysis of Average Prices by Brand:
The average price of mini vehicles is $10541.00
The average price of audi vehicles is $9212.00
The average price of mercedes_benz vehicles is $8536.00
The average price of bmw vehicles is $8261.00
The average price of skoda vehicles is $6353.00
The average price of hyundai vehicles is $5371.00
The average price of volkswagen vehicles is $5332.00
The average price of toyota vehicles is $5148.00
The average price of volvo vehicles is $4866.00
The average price of nissan vehicles is $4669.00
The average price of seat vehicles is $4315.00
The average price of mazda vehicles is $4059.00
The average price of citroen vehicles is $3761.00
The average price of ford vehicles is $3728.00
The average price of smart vehicles is $3518.00
The average price of mitsubishi vehicles is $3372.00
The average price of peugeot vehicles is $3065.00
The average price of opel vehicles is $2944.00
The average price of fiat vehicles is $2793.00
The average price of renault ve

In [48]:
# Storing Aggregate Data in a DataFrame
bmp_series = pd.Series(brand_mean_price)
pd.DataFrame(bmp_series, columns=['mean_price'])

mean_price
volkswagen           5332
opel                 2944
bmw                  8261
mercedes_benz        8536
audi                 9212
ford                 3728
renault              2431
peugeot              3065
fiat                 2793
seat                 4315
skoda                6353
nissan               4669
mazda                4059
smart                3518
citroen              3761
toyota               5148
hyundai              5371
volvo                4866
mini                10541
mitsubishi           3372

Observing the data, it's evident that the leading brands primarily originate from Germany. Volkswagen emerges as the most favored brand, offering vehicles within a moderate price range, averaging around $5332. BMW, Mercedes Benz, and Audi are also prominently featured, known for their higher price points.


<a id='pbb_alls'></a>
### Exploring Price by Brand for Entire Dataset

In [49]:
#Showing the price variations among ALL the brands
# Group by 'brand' and calculate the mean price 
price_by_brand = autos.groupby('brand')['price'].mean()
top_overall = price_by_brand.sort_values(ascending=False)
top_overall[:6]



brand
porsche       45624.271777
land_rover    18934.272727
jeep          11573.638889
jaguar        11525.554054
mini          10541.566986
audi           9212.930662
Name: price, dtype: float64

In [50]:
top_overall[-6:]

brand
lada        2647.724138
renault     2431.195699
trabant     1770.514706
daihatsu    1628.426230
rover       1586.492308
daewoo      1079.342105
Name: price, dtype: float64

Out data analysis shows that _Porsche_ takes the top spot for the most expensive brand, boasting an average price of $45,624. While _Land Rover_ follows behind with an average price of 18,934.

The top 3 cheapest brands are: _Daihatsu_, _Rover_, and _Daewoo_.

In [51]:
#mean mileage and mean price
top_20_brands


Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'volvo', 'mini', 'mitsubishi'],
      dtype='object')

<a id='miles'></a>
### Exploring Mileage
The next step is to explore mileage by brand.  We'll construct a dictionary comprising mileage data for the top 20 brands and subsequently convert it into a series.

In [52]:
brand_mileage = {}
for brand in top_20_brands:
    brand_only = autos[autos['brand'] == brand]
    mean_mileage = brand_only['odometer_km'].mean()
    brand_mileage[brand] = int(mean_mileage)
brand_mileage

#Convert both dictionaries to series objects
mean_mileage = pd.Series(brand_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_price).sort_values(ascending=False)



In [53]:
# Create a dataframe
brand_info = pd.DataFrame(mean_mileage, columns=['mean_mileage'])
brand_info

mean_mileage
volvo                138337
bmw                  132682
mercedes_benz        130796
audi                 129492
opel                 129383
volkswagen           128896
renault              128062
peugeot              127356
mitsubishi           125939
mazda                124871
ford                 124349
seat                 121641
citroen              119992
nissan               118711
fiat                 117567
toyota               116219
skoda                110993
hyundai              106718
smart                100511
mini                  88899

In [54]:
# Create a dataframe
brand_info['mean_price'] = mean_prices
brand_info

mean_mileage  mean_price
volvo                138337        4866
bmw                  132682        8261
mercedes_benz        130796        8536
audi                 129492        9212
opel                 129383        2944
volkswagen           128896        5332
renault              128062        2431
peugeot              127356        3065
mitsubishi           125939        3372
mazda                124871        4059
ford                 124349        3728
seat                 121641        4315
citroen              119992        3761
nissan               118711        4669
fiat                 117567        2793
toyota               116219        5148
skoda                110993        6353
hyundai              106718        5371
smart                100511        3518
mini                  88899       10541

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.

<a id='de_to_eng'></a>
### Replace German Words 
To enhance readability, we can substitute the German words in the dataset with their English equivalents.
Words to be replaced:

> `manuell` --> `manual`

>`automatik` --> `automatic`

>`benzin` --> `gas`

>`andere` --> `other`

>`elektro` --> `electric`


>`nein` --> `no`

>`ja` --> `yes`



In [55]:
autos['fuel_type'].unique()

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [56]:
autos['unrepaired_damage'].unique()

array(['nein', nan, 'ja'], dtype=object)

Using the map function, we will replace German words with English equivalents. 

In [57]:
#Create a dictionary of words to be replaced:
de_to_eng_map = {
    'manuell': 'manual',
    'automatik': 'automatic',
    'benzin': 'gas',
    'andere': 'other',
    'elektro': 'electric',
    'nein': 'no',
    'ja': 'yes'
}
autos['gearbox'] = autos['gearbox'].map(de_to_eng_map)
autos['fuel_type'] = autos['fuel_type'].map(de_to_eng_map)
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(de_to_eng_map)

In [58]:
autos.sample(5)

date_crawled                                               name  \
3340   2016-04-03 10:53:36                        Skoda_Oktavia_Ambiente_1_6l   
39295  2016-03-31 07:36:20                           Volkswagen_Polo_60_Servo   
1615   2016-03-18 17:36:24  Vw._Golf_1.9_TDI_Edition_Automatik_Gruene_Plakete   
37226  2016-03-14 15:57:20                                  Ford_Fiesta_Style   
33836  2016-03-24 20:00:43  Opel_Corsa_12V_City_KLIMA__2_JAHRE_TÜV__SCHIEB...   

        price   abtest vehicle_type  registration_year    gearbox  power_ps  \
3340   2700.0  control        kombi               2001     manual       102   
39295  1000.0     test   kleinwagen               1998     manual        82   
1615   2900.0     test   kleinwagen               2000  automatic        90   
37226   600.0  control   kleinwagen               1998     manual       102   
33836   990.0  control   kleinwagen               2000     manual        54   

         model  odometer_km  registration_month fuel_type       brand  \
3340   octavia     150000.0                   8       gas       skoda   
39295     polo     150000.0                  10       gas  volkswagen   
1615      golf     150000.0                  11       NaN  volkswagen   
37226   fiesta     150000.0                  11       gas        ford   
33836    corsa     125000.0                   8       gas        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
3340                 no  2016-04-03 00:00:00        39164  2016-04-07 11:46:30  
39295                no  2016-03-31 00:00:00        75177  2016-04-06 02:16:12  
1615                NaN  2016-03-18 00:00:00        71706  2016-04-05 22:44:36  
37226                no  2016-03-14 00:00:00        25761  2016-04-03 19:15:51  
33836               NaN  2016-03-24 00:00:00        40591  2016-03-29 06:46:14

<a id='uniform'></a>
### Convert the Dates to  Uniform Numeric Data

In [59]:
#Converting date to the datetime type and removing the time stamps.
# It essentially removes the time portion (hours, minutes, seconds) and keeps only the date information.

autos['date_crawled'] = pd.to_datetime(autos['date_crawled'])

#Date to integer
autos['date_integer'] = autos['date_crawled'].dt.strftime('%Y%m%d').astype(int)

#Reordering 'date_integer column'
autos
# autos['date_crawled'] = autos['date_crawled'].dt.date
# print(autos['date_crawled'].dtypes)
autos.head(3)


date_crawled                                        name   price  \
0 2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  5000.0   
1 2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  8500.0   
2 2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  8990.0   

    abtest vehicle_type  registration_year    gearbox  power_ps   model  \
0  control          bus               2004     manual       158  andere   
1  control    limousine               1997  automatic       286     7er   
2     test    limousine               2009     manual       102    golf   

   odometer_km  registration_month fuel_type       brand unrepaired_damage  \
0     150000.0                   3       NaN     peugeot                no   
1     150000.0                   6       gas         bmw                no   
2      70000.0                   7       gas  volkswagen                no   

            ad_created  postal_code            last_seen  date_integer  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54      20160326  
1  2016-04-04 00:00:00        71034  2016-04-06 14:45:08      20160404  
2  2016-03-26 00:00:00        35394  2016-04-06 20:15:37      20160326

The column with converted date _date_integer_ will appear in the end of dataset. We will not drop the original date column. 

In [60]:
#Converting date to the integer for 'last_seen' column
autos['last_seen'] = pd.to_datetime(autos['last_seen'])
print(autos['last_seen'].dtypes)

#Date to integer
autos['last_seen_integer'] = autos['last_seen'].dt.strftime('%Y%m%d').astype(int)
print(f'Data type of \"last_seen\" column: {autos["last_seen"].dtypes}')
print(f'Data type of \"last_seen_integer\" column: {autos["last_seen_integer"].dtypes}')

datetime64[ns]
Data type of "last_seen" column: datetime64[ns]
Data type of "last_seen_integer" column: int32


In [61]:
print(f'Data type of \"date_crawled\" column: {autos["date_crawled"].dtypes}')
print(f'Data type of \"date_integer\" column: {autos["date_integer"].dtypes}')

Data type of "date_crawled" column: datetime64[ns]
Data type of "date_integer" column: int32


In [62]:
#Converting date to the integer for 'ad_created' column
autos['ad_created'] = pd.to_datetime(autos['ad_created'])
print(autos['ad_created'].dtypes)

#Date to integer
autos['ad_created_integer'] = autos['ad_created'].dt.strftime('%Y%m%d').astype(int)

datetime64[ns]


In [63]:
autos.columns

Index(['date_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen', 'date_integer',
       'last_seen_integer', 'ad_created_integer'],
      dtype='object')

We can observe the new  date columns added to the dataset: _date_integer_, _last_seen_integer_, and _ad_created_integer_.
Rather than dropping columns, we can rearrange the positions of the new columns.

In [64]:
#Reordering 'date_integer column'
date_int_col = autos.pop('date_integer')
autos.insert(1, date_int_col.name, date_int_col)
autos.head(2)

date_crawled  date_integer  \
0 2016-03-26 17:47:46      20160326   
1 2016-04-04 13:38:56      20160404   

                                         name   price   abtest vehicle_type  \
0            Peugeot_807_160_NAVTECH_ON_BOARD  5000.0  control          bus   
1  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  8500.0  control    limousine   

   registration_year    gearbox  power_ps   model  odometer_km  \
0               2004     manual       158  andere     150000.0   
1               1997  automatic       286     7er     150000.0   

   registration_month fuel_type    brand unrepaired_damage ad_created  \
0                   3       NaN  peugeot                no 2016-03-26   
1                   6       gas      bmw                no 2016-04-04   

   postal_code           last_seen  last_seen_integer  ad_created_integer  
0        79588 2016-04-06 06:45:54           20160406            20160326  
1        71034 2016-04-06 14:45:08           20160406            20160404

<a id='common_b_m'></a>
### The Most Common Brand/Model Combinations


In [65]:
brand_model = autos.groupby(['brand', 'model'])['model'].count().sort_values(ascending=False).nlargest(10)
print('The top 10 Brand/Model Combinations\n')
print(brand_model)

The top 10 Brand/Model Combinations

brand          model   
volkswagen     golf        3900
bmw            3er         2686
volkswagen     polo        1688
opel           corsa       1681
               astra       1410
volkswagen     passat      1383
audi           a4          1256
mercedes_benz  c_klasse    1161
bmw            5er         1150
mercedes_benz  e_klasse     977
Name: model, dtype: int64


Given that Volkswagen is the most frequently occurring brand, it's unsurprising that it has the highest number of brand/model combinations.

<a id='miles_price'></a> 
### The Mileage/Price Analysis
We can split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.


In [66]:
autos['odometer_km'].value_counts()

150000.0    31245
125000.0     5022
100000.0     2069
90000.0      1719
80000.0      1401
70000.0      1198
60000.0      1137
50000.0       993
40000.0       798
5000.0        786
30000.0       758
20000.0       736
10000.0       232
Name: odometer_km, dtype: int64

We will create a custom function that will break mileage into groups. It will add a new column: _mileage_group_.

In [67]:
# Using .apply() method to divide odometer_km into groups based on the mileage.
def odometer_groups(odometer_km):
    if odometer_km < 30000:
        return '0-30k'
    elif odometer_km >= 30000 and odometer_km < 70000:
        return '30k-70k'
    elif odometer_km >= 70000 and odometer_km < 125000:
        return '70k-125k'
    else:
        return '125k-150k'
    
autos['mileage_group'] = autos['odometer_km'].apply(odometer_groups)
autos.head()

date_crawled  date_integer  \
0 2016-03-26 17:47:46      20160326   
1 2016-04-04 13:38:56      20160404   
2 2016-03-26 18:57:24      20160326   
3 2016-03-12 16:58:10      20160312   
4 2016-04-01 14:38:50      20160401   

                                                name   price   abtest  \
0                   Peugeot_807_160_NAVTECH_ON_BOARD  5000.0  control   
1         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  8500.0  control   
2                         Volkswagen_Golf_1.6_United  8990.0     test   
3  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  4350.0  control   
4  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  1350.0     test   

  vehicle_type  registration_year    gearbox  power_ps   model  ...  \
0          bus               2004     manual       158  andere  ...   
1    limousine               1997  automatic       286     7er  ...   
2    limousine               2009     manual       102    golf  ...   
3   kleinwagen               2007  automatic        71  fortwo  ...   
4        kombi               2003     manual         0   focus  ...   

   registration_month  fuel_type       brand unrepaired_damage ad_created  \
0                   3        NaN     peugeot                no 2016-03-26   
1                   6        gas         bmw                no 2016-04-04   
2                   7        gas  volkswagen                no 2016-03-26   
3                   6        gas       smart                no 2016-03-12   
4                   7        gas        ford                no 2016-04-01   

  postal_code           last_seen last_seen_integer  ad_created_integer  \
0       79588 2016-04-06 06:45:54          20160406            20160326   
1       71034 2016-04-06 14:45:08          20160406            20160404   
2       35394 2016-04-06 20:15:37          20160406            20160326   
3       33729 2016-03-15 03:16:28          20160315            20160312   
4       39218 2016-04-01 14:38:50          20160401            20160401   

   mileage_group  
0      125k-150k  
1      125k-150k  
2       70k-125k  
3       70k-125k  
4      125k-150k  

[5 rows x 21 columns]

In [68]:
#Unique values in mileage_group column
autos['mileage_group'].unique()

array(['125k-150k', '70k-125k', '30k-70k', '0-30k'], dtype=object)

In [69]:
avg_price = autos.groupby('mileage_group').agg({'price': 'mean'})

#Another way to group data based on milleage_group and average price
# avg_price = autos[['price','mileage_group']].groupby('mileage_group').mean(numeric_only=True)
avg_price

price
mileage_group              
0-30k          13902.068985
125k-150k       4041.391403
30k-70k        14149.298155
70k-125k        8950.401440

The analysis shows that cars with lower mileage are higher in price.

<a id='damage'></a>
###  The Price/Damage Correlation
How much cheaper are cars with damage than their non-damaged counterparts?

In [70]:
damage_price = autos.groupby('unrepaired_damage').agg({'price': 'mean'})
damage_price

price
unrepaired_damage             
no                 7010.737958
yes                2192.449527


The price of repaired or undamaged cars is nearly three times higher compared to unrepaired cars.